hack/generate-jobs.py

```diff
@@ -0,0 +1,145 @@
+    """
+    Generate job configuration files from one template.
+    Return list of errors.
+    """
+    config = configparser.ConfigParser()
+    config.read(fname)
+
+    template_name = config.get("DEFAULT", "template")
+    template_path = os.path.join(os.path.dirname(fname), template_name)
+    name = template_path.split(".")[0]
+    errors = []
+    with open(template_path, encoding="utf-8") as inp:
+        template = jinja2.Template(inp.read(), lstrip_blocks=True)
+        for kind in config.get("DEFAULT", "kinds").split(","):
+            tmp = tempfile.NamedTemporaryFile(
+                "w",
+                prefix=os.path.splitext(os.path.basename(__file__))[0],
+                delete=False,
+            )
+            with tmp:
+                header = (
+                    "# GENERATED FILE - DO NOT EDIT!\n#\n# "
+                    f"Instead, modify {template_name} and run `make generate-jobs`.\n"
+                )
+                for section in config.sections():
+                    tmp.write(
+                        template.render(
+                            config[section],
+                            kind=kind,
+                            job_name=f"{kind}-{section}",
+                            header=header,
+                        )
+                    )
+                    header = ""
+
+            out = f"{name}-{kind}.yaml"
+            if not os.path.exists(out):
+                if only_verify:
+                    os.unlink(tmp.name)
+                    errors.append(f"Can't verify content: {out} doesn't exist")
+                    continue
+            else:
+                equal = filecmp.cmp(out, tmp.name, shallow=False)
+                if only_verify:
+                    os.unlink(tmp.name)
+                    if not equal:
+                        errors.append(
+                            f"Generated content for {out} differs from existing"
+                        )
+                    continue
+                if equal:
+                    os.unlink(tmp.name)
+                    continue
+                if not overwrite:
+                    os.unlink(tmp.name)
+                    errors.append(
+                        f"Generated content for {out} differs from existing, "
+                        "use --owerwrite to update"
+                    )
+                    continue
+            shutil.move(tmp.name, out)
+
+    return errors
+
+
+def main(argv):
+    """Entry point."""
+    parser = argparse.ArgumentParser(
+        prog="Jobs Generator",
+        description="Generate job configuration files from templates",
+    )
+    parser.add_argument(
+        "pattern",
+        nargs="+",
+        help="config path pattern, e.g config/jobs/kubernetes/sig-node/*.conf",
+    )
+    parser.add_argument(
+        "--only-verify",
+        action="store_true",
+        help="Only verify if generated files are the same as existing",
+    )
```

Comment:
The way how "verify" is implemented right now only checks files that currently get generated.

Suppose someone changes a config such that some file no longer gets generated and then forgets to remove that generated file in the PR with that config change. That won't be caught.

What other verify scripts in Kubernetes do is to remove all generated files in the tree based on grepping and/or find. then re-generate them and look for diffs. But it's more complicated.

Perhaps the current approach is sufficient?

Reply:
I consider it sufficient at least for now. Let's see if it's going to be used anywhere else except DRA jobs.

Reply:
+1 to do what @pohly asked above as follow up.